nav2_velocity_smoother/README.md

```diff
@@ -0,0 +1,85 @@
+# Velocity Smoother
+
+The ``nav2_velocity_smoother`` is a package containing a lifecycle-component node for smoothing velocities sent by Nav2 to robot controllers.
+The aim of this package is to implement velocity, acceleration, and deadband smoothing from Nav2 to reduce wear-and-tear on robot motors and hardware controllers by smoothing out the accelerations/jerky movements that might be present with some local trajectory planners' control efforts.
+
+It supports differential drive and omnidirectional robot platforms primarily, but is applicable to ackermann as well with some intepretations of ``Twist``. It was built by [Steve Macenski](https://www.linkedin.com/in/steve-macenski-41a985101/) while at [Samsung Research](https://www.sra.samsung.com/). 
+
+See its [Configuration Guide Page](https://navigation.ros.org/configuration/packages/configuring-velocity-smoother.html) for additional parameter descriptions.
+
+## Features
+
+This package was created to do the following:
+
+- Limit velocity commands by kinematic constraints, including velocity and acceleration
+- Limit velocities based on deadband regions
+- Stop sending velocities after a given timeout duration of no new commands (due to stopped navigation)
+- Send a zero-velocity command at velocity timeout to stop the robot, in case not properly handled
+- Support Omni and differential drive robots (e.g. X, Y, Theta)
+- Smooth velocities proportionally in the same direction as commanded, whenever possible within kinematic limits
+- Provide open loop and closed loop options
+- Component nodes for use in single-process systems and stand-alone node format
+- Dynamically reconfigurable parameters
+
+## Design
+
+This is a lifecycle-component node, using the lifecycle manager for state management and composition for process management.
+It is designed to take in a command from Nav2's controller server and smooth it for use on robot hardware controllers.
+Thusly, it takes in a command via the `cmd_vel` topic and produces a smoothed output on `smoothed_cmd_vel`.
+
+The node is designed on a regular timer running at a configurable rate.
```

Comment:
I'm not sure what you mean by `speed_limit`? You mean with the speed restricted layer? Sure, that can change, and these parameters are dynamically reconfigurable.

Though, if you had a limit of `1.0` and in your zone you have a limit of `0.5`, so your `cmd` coming in is `0.5`, the velocity smoother will still smooth the speed down to `0.5` even if it doesn't specifically know why that command is being sent. Even if this node had awareness, it would still enforce kinematic constraints so there would be the same ramp down profile from `1.0` to `0.5` -- whether it comes from the reflected `cmd`'s or explicit notification of the change

Reply:
Sure  when `speed_limit` is < `smoother.max_vel.vx` everything is ok as you describe.

I am really talking about the case of `speed_limit` < `smoother.min_vel.vx` which means that the smoother is causing a velocity that is in fact _higher_ than the prescribed speed limit.  

To put numbers to it:
limit of `0.5` but the smoother min velocity is `1.0` -- the smoother would be violating the speed limit in this case.

Maybe it's just a sign of a configured system, so maybe it doesn't need any further clarification. Just trying to imagine how things might go wrong when trying to use this.

Reply:
That is an interesting problem I had not considered, but `min_velocity: [-0.5, 0.0, -2.5]` is signed, so either `0.0` if you don't want to reverse, or the maximum negative velocity if reversing. So I don't think that situation is actually possible. Although that was not an intentional design decision, I just got lucky :wink: